compiler/riscv/codegen/RVInstruction.hpp

```diff
@@ -1298,6 +1298,65 @@ class AdminInstruction : public TR::Instruction
    virtual int32_t estimateBinaryLength(int32_t currentEstimate);
    };
 
+class DataInstruction : public TR::Instruction
+   {
+   uint32_t _sourceImmediate;
+
+public:
+
+   /*
+    * @brief Constructor
+    * @param[in] op : instruction opcode
+    * @param[in] node : node
+    * @param[in] imm : immediate value
+    * @param[in] cg : CodeGenerator
+    */
+   DataInstruction(TR::InstOpCode::Mnemonic op, TR::Node *node, uint32_t imm, TR::CodeGenerator *cg)
+      : TR::Instruction(op, node, cg), _sourceImmediate(imm)
+      {
+      TR_ASSERT(op == TR::InstOpCode::dd, "Opcode of data instruction is not dd");
+      }
+   /*
+    * @brief Constructor
+    * @param[in] op : instruction opcode
+    * @param[in] node : node
+    * @param[in] imm : immediate value
+    * @param[in] precedingInstruction : preceding instruction
+    * @param[in] cg : CodeGenerator
+    */
+   DataInstruction(TR::InstOpCode::Mnemonic op, TR::Node *node, uint32_t imm,
+                       TR::Instruction *precedingInstruction, TR::CodeGenerator *cg)
+      : TR::Instruction(op, node, precedingInstruction, cg), _sourceImmediate(imm)
+      {
+      TR_ASSERT(op == TR::InstOpCode::dd, "Opcode of data instruction is not dd");
+      }
+
+   /**
+    * @brief Gets instruction kind
+    * @return instruction kind
+    */
+   virtual Kind getKind() { return IsData; }
+
+   /**
+    * @brief Gets source immediate
+    * @return source immediate
+    */
+   uint32_t getSourceImmediate() {return _sourceImmediate;}
+   /**
+    * @brief Sets source immediate
+    * @param[in] si : immediate value
+    * @return source immediate
+    */
+   uint32_t setSourceImmediate(uint32_t si) {return (_sourceImmediate = si);}
```

Comment:
Do we forsee a use of this API? If we can guarantee immutability of instructions we will be better off.

Reply:
Yes, there are users:
1. https://github.com/eclipse/omr/blob/0f81a4f102f17c456770dc12944e266c08907961/compiler/riscv/codegen/OMRCodeGenerator.cpp#L150-L156

2. In OpenJ9, in private linkage there's (will be) code like:
    ```
   void J9::RV::PrivateLinkage::performPostBinaryEncoding()
   {
   // --------------------------------------------------------------------------
   // Encode the size of the interpreter entry area into the linkage info word
   //
   TR_ASSERT_FATAL(cg()->getReturnTypeInfoInstruction(),
                   "Expecting the return type info instruction to be created");

   auto linkageInfoWordInstruction = cg()->getReturnTypeInfoInstruction();
   uint32_t linkageInfoWord = linkageInfoWordInstruction->getSourceImmediate();

   intptr_t jittedMethodEntryAddress = reinterpret_cast<intptr_t>(getJittedMethodEntryPoint()->getBinaryEncoding());
   intptr_t interpretedMethodEntryAddress = reinterpret_cast<intptr_t>(getInterpretedMethodEntryPoint()->getBinaryEncoding());

   linkageInfoWord = (static_cast<uint32_t>(jittedMethodEntryAddress - interpretedMethodEntryAddress) << 16) | linkageInfoWord;
   linkageInfoWordInstruction->setSourceImmediate(linkageInfoWord);

   *(uint32_t *)(linkageInfoWordInstruction->getBinaryEncoding()) = linkageInfoWord;

   // Set recompilation info
    ```
    This pretty much copies AArch64 implementation, see https://github.com/eclipse/openj9/blob/56b263b768417119ed12dec405cc2e5803962b5f/runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp#L1569